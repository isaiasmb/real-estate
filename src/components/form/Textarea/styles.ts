import styled, { css, DefaultTheme } from 'styled-components'

import { TextareaProps } from '.'

type WrapperProps = Pick<TextareaProps, 'disabled'> & { error?: boolean }

export const TextareaWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    background-color: ${theme.colors.white};
    border-radius: 0.2rem;
    border: 0.1rem solid;
    border-color: ${theme.colors.lightGray};
    padding: 0 ${theme.spacings.xsmall};
    margin: ${theme.spacings.xxxsmall} 0;
    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`

export const Textarea = styled.textarea`
  ${({ theme }) => css`
    width: 100%;
    resize: none;
    outline: none;
    background: transparent;
    border: 0;
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    cursor: pointer;
  `}
`

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.red};
    font-size: ${theme.font.sizes.xsmall};
  `}
`

const wrapperModifiers = {
  error: (theme: DefaultTheme) => css`
    ${Label} {
      color: ${theme.colors.red};
    }
  `,
  disabled: (theme: DefaultTheme) => css`
    ${Label} {
      cursor: not-allowed;
      color: ${theme.colors.gray};
      &::placeholder {
        color: currentColor;
      }
    }
  `
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, error, disabled }) => css`
    ${error && wrapperModifiers.error(theme)}
    ${disabled && wrapperModifiers.disabled(theme)}
  `}
`
