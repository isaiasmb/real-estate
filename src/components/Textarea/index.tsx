import * as S from './styles'

const Textarea = () => (
  <S.Wrapper>
    <h1>Textarea</h1>
  </S.Wrapper>
)

export default Textarea
