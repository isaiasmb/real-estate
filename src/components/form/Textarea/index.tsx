import { useState, TextareaHTMLAttributes } from 'react'

import * as S from './styles'

export type TextareaProps = {
  onInput?: (value: string) => void
  label?: string
  labelFor?: string
  initialValue?: string
  disabled?: boolean
  error?: string
} & TextareaHTMLAttributes<HTMLTextAreaElement>

const Textarea = ({
  label,
  labelFor = '',
  initialValue = '',
  disabled = false,
  error,
  onInput,
  rows = 5,
  ...props
}: TextareaProps) => {
  const [value, setValue] = useState(initialValue)

  const onChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    const newValue = e.currentTarget.value
    setValue(newValue)

    !!onInput && onInput(newValue)
  }

  return (
    <S.Wrapper disabled={disabled} error={!!error}>
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}
      <S.TextareaWrapper>
        <S.Textarea
          onChange={onChange}
          value={value}
          disabled={disabled}
          rows={rows}
          {...props}
        />
      </S.TextareaWrapper>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  )
}

export default Textarea
