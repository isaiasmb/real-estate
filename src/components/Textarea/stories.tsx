import { Story, Meta } from '@storybook/react/types-6-0'
import Textarea from '.'

export default {
  title: 'Textarea',
  component: Textarea
} as Meta

export const Default: Story = () => <Textarea />
