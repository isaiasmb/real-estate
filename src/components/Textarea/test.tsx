import { render, screen } from '@testing-library/react'

import Textarea from '.'

describe('<Textarea />', () => {
  it('should render the heading', () => {
    const { container } = render(<Textarea />)
    expect(
      screen.getByRole('heading', { name: /Textarea/i })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
