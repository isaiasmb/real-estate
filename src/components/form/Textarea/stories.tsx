import { Story, Meta } from '@storybook/react/types-6-0'

import Textarea, { TextareaProps } from '.'

export default {
  title: 'Form/Textarea',
  component: Textarea,
  args: {
    label: 'Description',
    labelFor: 'description',
    id: 'description',
    initialValue: '',
    placeholder: 'Type the description'
  },
  argTypes: {
    onInput: { action: 'changed' },
    icon: { type: '' }
  }
} as Meta

export const Default: Story<TextareaProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <Textarea {...args} />
  </div>
)

export const WithError: Story<TextareaProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <Textarea {...args} />
  </div>
)

WithError.args = {
  error: 'Ops...something is wrong'
}
