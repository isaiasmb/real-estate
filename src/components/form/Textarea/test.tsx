import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { renderWithTheme } from 'utils/tests/helpers'

import Textarea from '.'

describe('<Textarea />', () => {
  it('Renders with Label', () => {
    renderWithTheme(<Textarea label="Label" labelFor="Field" id="Field" />)

    expect(screen.getByLabelText('Label')).toBeInTheDocument()
  })

  it('Renders without Label', () => {
    renderWithTheme(<Textarea />)

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('Renders with placeholder', () => {
    renderWithTheme(<Textarea placeholder="hey you" />)

    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument()
  })

  it('Changes its value when typing', async () => {
    const onInput = jest.fn()
    renderWithTheme(
      <Textarea
        onInput={onInput}
        label="Textarea"
        labelFor="Textarea"
        id="Textarea"
      />
    )

    const textarea = screen.getByRole('textbox')
    const text = 'This is my new text'
    userEvent.type(textarea, text)

    await waitFor(() => {
      expect(textarea).toHaveValue(text)
      expect(onInput).toHaveBeenCalledTimes(text.length)
    })
    expect(onInput).toHaveBeenCalledWith(text)
  })

  it('Does not changes its value when disabled', async () => {
    const onInput = jest.fn()
    renderWithTheme(
      <Textarea
        onInput={onInput}
        label="Textarea"
        labelFor="Textarea"
        id="Textarea"
        disabled
      />
    )

    const textarea = screen.getByRole('textbox')
    expect(textarea).toBeDisabled()

    const text = 'This is my new text'
    userEvent.type(textarea, text)

    await waitFor(() => {
      expect(textarea).not.toHaveValue(text)
    })
    expect(onInput).not.toHaveBeenCalled()
  })

  it('Renders with error', () => {
    const { container } = renderWithTheme(
      <Textarea label="Textarea" labelFor="Textarea" error="Error message" />
    )

    expect(screen.getByText('Error message')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Is accessible by tab', () => {
    renderWithTheme(
      <Textarea label="Textarea" labelFor="Textarea" id="Textarea" />
    )

    const textarea = screen.getByLabelText('Textarea')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(textarea).toHaveFocus()
  })
})

it('Is not accessible by tab when disabled', () => {
  renderWithTheme(
    <Textarea label="Textarea" labelFor="Textarea" id="Textarea" disabled />
  )

  const textarea = screen.getByLabelText('Textarea')
  expect(document.body).toHaveFocus()

  userEvent.tab()
  expect(textarea).not.toHaveFocus()
})
