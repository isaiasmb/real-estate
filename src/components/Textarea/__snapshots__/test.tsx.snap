// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Textarea /> should render the heading 1`] = `
<main
  class=""
>
  <h1>
    Textarea
  </h1>
</main>
`
